src/parsers/espresso/7.2/parser.js

```diff
@@ -0,0 +1,292 @@
+ * @return {{vectors: Number[][], units: String}}
+ */
+function getCellConfig(text) {
+    const match = regex.cellParameters.exec(text);
+    if (match) {
+        const units = match[1];
+        const values = match.slice(2, 11);
+        // creating 3 vectors from 9 values
+        const vectors = Array.from({ length: 3 }, (_, i) =>
+            values.slice(i * 3, i * 3 + 3).map(Number),
+        );
+        return { vectors, units };
+    }
+    throw new Error("Couldn't read cell parameters");
+}
+
+/**
+ * @summary Return units and scaling factor according to Quantum Espresso docs
+ * @param {String} units - units from ATOMIC_POSITIONS card
+ * @returns {{_units: String, scalingFactor: Number}}
+ */
+function convertFromEspressoUnits(units) {
+    let _units, scalingFactor;
+    switch (units) {
+        case "alat":
+            scalingFactor = 1.0;
+            _units = ATOMIC_COORD_UNITS.crystal;
+            break;
+        case "bohr":
+            scalingFactor = coefficients.BOHR_TO_ANGSTROM;
+            _units = ATOMIC_COORD_UNITS.cartesian;
+            break;
+        case "angstrom":
+            scalingFactor = 1.0;
+            _units = ATOMIC_COORD_UNITS.cartesian;
+            break;
+        case "crystal":
+            scalingFactor = 1.0;
+            _units = ATOMIC_COORD_UNITS.crystal;
+            break;
+        case "crystal_sg":
+            throw new Error("crystal_sg not supported yet");
+        default:
+            throw new Error(`Units ${units} not supported`);
+    }
+    return { _units, scalingFactor };
+}
+
+/**
+ * @summary Read atomic positions from ATOMIC_POSITIONS card
+ * @param {String} text - cards data
+ * @returns {{elements: Object[], coordinates: Object[], constraints: Object[], units: String}}
+ */
+function getAtomicPositions(text) {
+    const atomicSpeciesMatches = Array.from(text.matchAll(regex.atomicSpecies));
+    // eslint-disable-next-line no-unused-vars
+    const atomicSpecies = atomicSpeciesMatches.map((match) => ({
+        element: match[1],
+        mass: parseFloat(match[2]),
+        potential: match[3],
+    }));
+    const atomicPositionsMatches = Array.from(text.matchAll(regex.atomicPositions));
+    const units = text.match(regex.atomicPositionsUnits)[1];
+    const { _units, scalingFactor } = convertFromEspressoUnits(units);
+
+    const elements = atomicPositionsMatches.map((match, index) => ({
+        id: index,
+        value: match[1],
+    }));
+    const coordinates = atomicPositionsMatches.map((match, index) => ({
+        id: index,
+        value: [
+            parseFloat(match[2]) * scalingFactor,
+            parseFloat(match[3]) * scalingFactor,
+            parseFloat(match[4]) * scalingFactor,
+        ],
+    }));
+    const constraints = atomicPositionsMatches
+        .filter((match) => match[5] && match[6] && match[7]) // Check if all three constraints exist
+        .map((match, index) => ({
+            id: index,
```

Comment:
If only a subset of coordinates has constraints, the indices will not be different due to the filtering in line 147. Perhaps we can use `.reduce()` here.

Reply:
Ok, will change. I thought constraints either present on all atoms or not at all